File: events/year_2024/day_03/day_03.py
"""
Advent of Code - Year 2024 - Day 3
https://adventofcode.com/2024/day/3
"""

from collections.abc import Iterator
import re


def solver(program: str) -> Iterator[int]:
    """
    For the first part, scan the program for uncorrupted mul operations. Then remove all
    don't()...do() enclosure and remove the last disabled instructions following a don't()
    """
    yield scan_mul(program)

    program = re.sub(r"don\'t\(\)([\s\S]*?)do\(\)", "", program)
    program = program.split("don\'t()", 1)[0]
    yield scan_mul(program)


def scan_mul(program: str) -> int:
    """
    Use regex expression to extract uncorrupted instructions with one up to 3 digits. Then add the
    product of matched pairs and return it
    """
    score = 0
    uncorrupted = re.findall(r"mul\((\d{1,3}),(\d{1,3})\)", program)
    for a, b in uncorrupted:
        score += int(a) * int(b)
    return score

File: events/year_2024/day_03/day_03.rs
aoc::main!();

fn preprocessing(puzzle_input: &str) -> &str {
    puzzle_input
}

fn part_1(program: &str) -> u32 {
    scan_mul(program)
}

fn part_2(program: &str) -> u32 {
    scan_mul(
        Regex::new(r"don't\(\)([\s\S]*?)do\(\)")
            .unwrap()
            .replace_all(program, "")
            .split_once("don't()")
            .unwrap()
            .0,
    )
}

fn scan_mul(program: &str) -> u32 {
    Regex::new(r"mul\((\d{1,3}?),(\d{1,3}?)\)")
        .unwrap()
        .captures_iter(program)
        .map(|pair| pair.extract())
        .map(|(_, [a, b])| a.parse::<u32>().unwrap() * b.parse::<u32>().unwrap())
        .sum()
}
